src/CharterManager.sol

```diff
@@ -0,0 +1,257 @@
+    function dai(address) external view returns (uint256);
+    function fork(bytes32, address, address, int256, int256) external;
+    function frob(bytes32, address, address, address, int256, int256) external;
+    function flux(bytes32, address, address, uint256) external;
+    function move(address, address, uint256) external;
+    function hope(address) external;
+    function nope(address) external;
+    function ilks(bytes32) external view returns (
+        uint256 Art,  // [wad]
+        uint256 rate, // [ray]
+        uint256 spot, // [ray]
+        uint256 line, // [rad]
+        uint256 dust  // [rad]
+    );
+}
+
+interface ManagedGemJoinLike {
+    function gem() external view returns (address);
+    function ilk() external view returns (bytes32);
+    function join(address, uint256) external;
+    function exit(address, address, uint256) external;
+}
+
+interface TokenLike {
+    function approve(address, uint256) external;
+    function transferFrom(address, address, uint256) external;
+}
+
+contract UrnProxy {
+    address immutable public usr;
+
+    constructor(address vat_, address usr_) public {
+        usr = usr_;
+        VatLike(vat_).hope(msg.sender);
+    }
+}
+
+contract CharterManager {
+    mapping (address => uint256) public wards;
+    address public implementation;
+
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event SetImplementation(address indexed);
+
+    constructor() public {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(msg.sender);
+    }
+
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(msg.sender);
+    }
+
+    modifier auth {
+        require(wards[msg.sender] == 1, "CharterManager/non-authed");
+        _;
+    }
+
+    function setImplementation(address implementation_) external auth {
+        implementation = implementation_;
+        emit SetImplementation(implementation_);
+    }
+
+    fallback() external {
+        address _impl = implementation;
+        require(_impl != address(0));
+
+        assembly {
+            let ptr := mload(0x40)
+            calldatacopy(ptr, 0, calldatasize())
+            let result := delegatecall(gas(), _impl, ptr, calldatasize(), 0, 0)
+            let size := returndatasize()
+            returndatacopy(ptr, 0, size)
+
+            switch result
+            case 0 { revert(ptr, size) }
+            default { return(ptr, size) }
+        }
+    }
+}
+
+contract CharterManagerImp {
+    // --- Data ---
+    mapping (address => uint256) public wards;
+    bytes32 slot1;
+    mapping (address => address) public proxy; // UrnProxy per user
+    mapping (address => mapping (address => uint256)) public can;
+
+    mapping (bytes32 => bool)                        public gate; // allow only permissioned vaults
```

Comment:
I'd use `uint256` instead of `bool` here to avoid the bitmasking gas costs. `bool` is really only justified when you're able to pack multiple variables into a storage slot.

Reply:
Done.